application/src/main/kotlin/net/corda/v5/application/messaging/FlowMessaging.kt

```diff
@@ -8,12 +9,38 @@ import net.corda.v5.base.types.MemberX500Name
 interface FlowMessaging {
 
     /**
-     * Creates a communication session with [party]. Subsequently you may send/receive using this session object. Note
-     * that this function does not communicate in itself, the counter-flow will be kicked off by the first send/receive.
+     * Creates a communication session with another member. Subsequently you may send/receive using this session object.
+     * Note that this function does not communicate in itself, the counter-flow will be kicked off by the first
+     * send/receive.
+     *
+     * Initiated flows are initiated with context based on the context of the initiating flow. The context of the
+     * initiating flow is snapshotted by the returned session when this method is called. Altering the parent context
+     * has no effect on the context of the session after this point, and therefore it has no effect on the context of
+     * the initiated flow either.
+     *
+     * @param x500Name The X500 name of the member to communicate with.
+     * @return The session.
      */
     @Suspendable
     fun initiateFlow(x500Name: MemberX500Name): FlowSession
 
+    /**
+     * Creates a communication session with another member. Subsequently you may send/receive using this session object.
+     * Note that this function does not communicate in itself, the counter-flow will be kicked off by the first
+     * send/receive.
+     *
+     * This overload takes a mutator of context properties. Any properties set or modified against the context passed to
+     * this mutator will be propagated to initiated flows and all that flow's initiated flows and sub flows down the
+     * stack. The properties passed to the mutator are pre-populated with the parent context properties, see
+     * [FlowContextProperties].
+     *
+     * @param x500Name The X500 name of the member to communicate with.
+     * @param flowContextPropertiesMutator A mutator of context properties
+     * @return The session.
+     */
+    @Suspendable
+    fun initiateFlow(x500Name: MemberX500Name, flowContextPropertiesMutator: FlowContextPropertiesMutator): FlowSession
```

Comment:
Should update the FlowMessagingJavaApiTest for this as well.

Reply:
Is this overcomplicating things? When exactly is the mutator applied? Is it applied immediately, or is it applied at first packet when we start handshake? If applied immediately should we just pass in sessionContext: Map<String, String> instead?

Reply:
Also I am trying to think what happens when we create a session in an outer flow and then pass the session down into an inline subflow. Is the base session context derived from the original non-inline subflow, or from the point of use?

Reply:
This was also considered, but ruled out because
* then you present two different apis to the user to set context
* you don't get any of the validation you get from FlowContext
* when used with a lambda in Java or Kotlin the api is trivial to use:

```
In Java this can be passed a lambda, e.g.
 session = flowMessaging.initiateFlow(vnName, (flowContextProperties) -> {
                // mutate context
              });

In Kotlin:
        val session =
            flowMessaging.initiateFlow(flowEngine.virtualNodeName) { flowContextProperties ->
              // mutate context
            }
```
No user created maps need to be involved in the process.

I attempted to document that the properties are snapshotted here: https://github.com/corda/corda-api/pull/542/files#diff-143329fd07352dd773d710271bd76e65b232e150ff0788873c3573b8a0ecb901R34
perhaps that isn't clear enough? On re-reading I see it's probably clearer on the other overload.

Reply:
Re: subflows. Yes, the session context remains immutable, even if you pass it to a subflow, so changing the subflow context would then have no effect at all. Whilst this version prohibits that kind of "advanced" context capturing, the previous iteration captured context at the first point you sent a message and then, no matter what you did, it stayed the same, even though that wasn't quite so apparent to the user because they could continue to modify it.

So this is limiting the feature a little, but it always works the same way and results shouldn't be unexpected.